KAFKA-16523; kafka-metadata-quorum: support add-controller and remove-controller

cmccabe opened the pull request:
This PR adds support for add-controller and remove-controller in the kafka-metadata-quorum.sh command-line tool. It also fixes some minor server-side bugs that blocked the tool from working.

In kafka-metadata-quorum.sh, the implementation of remove-controller is fairly straightforward. It just takes some command-line flags and uses them to invoke AdminClient. The add-controller implementation is a bit more complex because we have to look at the new controller's configuration file. The parsing logic for the `advertised.listeners` and `listeners` server configurations that we need was previously implemented in the :core module. However, the gradle module where kafka-metadata-quorum.sh lives, :tools, cannot depend on :core. Therefore, I moved listener parsing into `SocketServerConfigs.listenerListToEndPoints`. This will be a small step forward in our efforts to move Kafka configuration out of :core.

I also made some minor changes in kafka-metadata-quorum.sh and Kafka-storage-tool.sh to handle --help without displaying a backtrace on the screen, and give slightly better error messages on stderr. Also, in DynamicVoter.toString, we now enclose the host in brackets if it contains a colon (as IPV6 addresses can).

This PR fixes our handling of clusterId in `addRaftVoter` and `removeRaftVoter`, in two ways. Firstly, it marks `clusterId` as nullable in the `AddRaftVoterRequest.json` and `RemoveRaftVoterRequest.json` schemas, as it was always intended to be. Secondly, it allows AdminClient to optionally send clusterId, by using `AddRaftVoterOptions` and `RemoveRaftVoterOptions`. We now also remember to properly set `timeoutMs` in `AddRaftVoterRequest`. This PR adds unit tests for `KafkaAdminClient#addRaftVoter` and `KafkaAdminClient#removeRaftVoter`, to make sure they are sending the right things.

Finally, I fixed some minor server-side bugs that were blocking the handling of these RPCs. Firstly, `ApiKeys.ADD_RAFT_VOTER` and `ApiKeys.REMOVE_RAFT_VOTER` are now marked as forwardable so that forwarding from the broker to the active controller works correctly. Secondly, `org.apache.kafka.raft.KafkaNetworkChannel` has now been updated to enable `API_VERSIONS_REQUEST` and `API_VERSIONS_RESPONSE`.

Co-authored-by: Murali Basani <muralidhar.basani@aiven.io>

cmccabe commented:
@muralibasani @jsancio this is a version of @muralibasani 's PR that doesn't require adding the dependency that we were concerned about earlier, between core and tools. Please check it out.

ahuang98 commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -158,6 +170,49 @@ private static void addDescribeSubParser(ArgumentParser parser) {
+
+        addControllerParser
+            .addArgument("--controller-config")
+            .help("The configuration of the controller that we are removing.")
```
`that we are adding`

cmccabe committed:
KAFKA-16523: kafka-metadata-quorum: support add-controller and remove-controller

In AddRaftVoterRequest.json and RemoveRaftVoterRequest.json, make ClusterId nullable, since we
don't always set it.

jsancio commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4667,6 +4667,9 @@ AddRaftVoterRequest.Builder createRequest(int timeoutMs) {
                         setPort(endpoint.port())));
                 return new AddRaftVoterRequest.Builder(
                    new AddRaftVoterRequestData().
+                       setClusterId(options.clusterId().isPresent() ?
+                           options.clusterId().get() : null).
```
This is the same as
```java
                       setClusterId(options.clusterId().orElse(null)).
```

cmccabe replied:
ok

jsancio commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4711,6 +4714,8 @@ public RemoveRaftVoterResult removeRaftVoter(
             RemoveRaftVoterRequest.Builder createRequest(int timeoutMs) {
                 return new RemoveRaftVoterRequest.Builder(
                     new RemoveRaftVoterRequestData().
+                        setClusterId(options.clusterId().isPresent() ?
+                            options.clusterId().get() : null).
```
This is the same as
```java
                       setClusterId(options.clusterId().orElse(null)).
```

cmccabe replied:
ok

jsancio commented on server/src/test/java/org/apache/kafka/network/SocketServerConfigsTest.java:
```diff
@@ -0,0 +1,107 @@
+    @Test
+    public void testAnotherListenerListToEndPointsWithIpV6() {
+        assertEquals(Arrays.asList(
+            new Endpoint("SASL_SSL", SecurityProtocol.SASL_SSL, "fe80::b1da:69ca:57f7:63d8%3", 9092)),
+                SocketServerConfigs.listenerListToEndPoints("SASL_SSL://[fe80::b1da:69ca:57f7:63d8%3]:9092",
+                    SocketServerConfigs.DEFAULT_NAME_TO_SECURITY_PROTO));
```
Interesting that the character `%` is allowed in IPv6: https://superuser.com/a/99753

cmccabe replied:
Yeah, it's a puzzling design decision, especially given that  `%` is not a URL-safe character.

jsancio commented on tools/src/main/java/org/apache/kafka/tools/MetadataQuorumCommand.java:
```diff
@@ -291,4 +320,179 @@ public String toString() {
+                    listenerName);
+            }
+            results.add(new RaftVoterEndpoint(endpoint.listenerName().get(),
+                    endpoint.host() == null ? "localhost" : endpoint.host(),
```
Should it be an error if the host is `null`? Do we assume localhost for brokers if they don't specify a host in the listeners or advertise listeners configuration?

cmccabe replied:
I was on the fence about this.

I think for local testing, I can definitely imagine some people putting `CONTROLLER://:9093` in the advertised listeners, and not thinking anything more about it. This would certainly be a bad idea in production, though.

We could throw an exception in that situation, or we could translate it to `localhost`... Perhaps the exception is better?

jsancio reviewed (commented):
Thanks for the changes they look good to me. Just a question and some minor suggestions.

cmccabe committed:
Address review comments, nicer formatting for --help, set ApiKeys properly.

cmccabe committed:
Merge branch 'trunk' into KAFKA-16523-III

cmccabe committed:
fix checkstyle

ahuang98 commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -8232,4 +8242,92 @@ boolean callHasExpired(KafkaAdminClient.Call call) {
+            if (fail) {
+                TestUtils.assertFutureThrows(result.all(), Errors.DUPLICATE_VOTER.exception().getClass());
+            } else {
+                result.all().get();
```
nit: should we assertNull? would clarify to readers of this code that we expect the result of a successful add voter to be null

ahuang98 commented on core/src/main/scala/kafka/tools/StorageTool.scala:
```diff
@@ -190,8 +193,8 @@ object StorageTool extends Logging {
-      help("A list of controller quorum voter ids, directories, and hostname:port pairs. The same values must be used to format all nodes. For example:\n" +
-        "0@localhost:8082:JEXY6aqzQY-32P5TStzaFg@,1@localhost:8083:MvDxzVmcRsaTz33bUuRU6A,2@localhost:8084:07R5amHmR32VDA6jHkGbTA\n").
+      help("The initial controllers, as a comma-separated list of id@hostname:port:directory. The same values must be used to format all nodes. For example:\n" +
+        "0@example.com:8082:JEXY6aqzQY-32P5TStzaFg,1@example.com:8083:MvDxzVmcRsaTz33bUuRU6A,2@example.com:8084:07R5amHmR32VDA6jHkGbTA\n").
```
should we an include an ipv6 example?

jsancio replied:
Good point. We need to mentioned somewhere that IPv6 need to be escaped with the `[]` characters.

ahuang98 commented on raft/src/main/java/org/apache/kafka/raft/KafkaNetworkChannel.java:
```diff
@@ -185,6 +187,10 @@ static AbstractRequest.Builder<? extends AbstractRequest> buildRequest(ApiMessag
+        if (requestData instanceof ApiVersionsRequestData)
+            return new ApiVersionsRequest.Builder((ApiVersionsRequestData) requestData,
+                ApiKeys.API_VERSIONS.oldestVersion(),
+                ApiKeys.API_VERSIONS.latestVersion());
```
why do we need to specify these versions explicitly? is it because we need `enableUnstableLastVersion==true`?

cmccabe replied:
The reason is more mundane than that. I didn't want to have a lot of constructors, so I decided to have:

- no-args
- required-version
- data, min-version, max-version

I could create lots of permutations, like:

- no-args 
- required-version
-  data
- data, required-version
- data, min-version, max-version

but 3 constructors seems like enough?

jsancio commented on raft/src/main/java/org/apache/kafka/raft/RaftUtil.java:
```diff
@@ -67,6 +68,8 @@ public static ApiMessage errorResponse(ApiKeys apiKey, Errors error) {
             case FETCH_SNAPSHOT:
                 return new FetchSnapshotResponseData().setErrorCode(error.code());
+            case API_VERSIONS:
+                return new ApiVersionsResponseData().setErrorCode(error.code());
```
I need to add UpdateVoter to this switch too. Let me do that in my PR.

jsancio reviewed (commented):
Thanks for the changes. The previous build had failures in `KafkaConfigTest` were those fix this the latest code?

cmccabe commented:
> Thanks for the changes. The previous build had failures in KafkaConfigTest were those fix this the latest code?

yes.

> Good point. We need to mentioned somewhere that IPv6 need to be escaped with the [] characters.

This might be somewhere in the docs. File a follow-on JIRA to look?

jsancio commented:
There are some failures in the command line tests and the build didn't finish. Triggering another build.

cmccabe committed:
Fix tests

cmccabe commented:
> There are some failures in the command line tests and the build didn't finish. Triggering another build

I fixed two unit tests and pushed again.

cmccabe committed:
Merge branch 'trunk' into KAFKA-16523

cmccabe committed:
fix checkstyle

jsancio reviewed (approved):
LGTM if the build is green.
